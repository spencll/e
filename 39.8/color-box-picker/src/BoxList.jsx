import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";

// ***BoxList***
// Place your state that contains all of the boxes here. This component should render all of the ***Box*** components along with the ***NewBoxForm*** component

function BoxList() {
  const INITIAL_STATE = [];

  //state
  const [boxes, setBoxes] = useState(INITIAL_STATE);

  //adding box, appending new box object to array of boxes with random id for key
  const addBox = (newBox) => {
    setBoxes((boxes) => [...boxes, { ...newBox, id: uuid() }]);
  };

  const deleteBox = (id) => {
    setBoxes((prevBoxes) => prevBoxes.filter((box) => box.id !== id));
  };

  return (
    <div>
      <h3>Boxes</h3>
      {/* Render new form and passing function down to form*/}
      <NewBoxForm addBox={addBox} />

      {/* Extracting properties from each box object to generate new box */}
      <div>
        {boxes.map(({ id, bgColor, height, width }) => (
          <Box
            id={id}
            key={id}
            bgColor={bgColor}
            height={parseInt(height)}
            width={parseInt(width)}
            onDelete={deleteBox}
          />
        ))}
      </div>
    </div>
  );
}

export default BoxList;
