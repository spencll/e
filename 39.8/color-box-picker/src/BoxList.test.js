import React from "react";
import { render, fireEvent} from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new box and delete box after", function () {
  //Getting search functions from render
  const { queryByText, getByLabelText} = render(<BoxList />);

  //Geting targets
  const input = getByLabelText("Color:");
  const btn = queryByText("Add box");

  //Add green into color input and submit
  fireEvent.change(input, { target: { value: "green" } });
  fireEvent.click(btn);
  let newBox = queryByText("Here is box");
  expect(newBox).toBeInTheDocument();
//Targeting delete button and firing delete 
  const deleteBtn= queryByText("Delete")
  fireEvent.click(deleteBtn)
  expect(newBox).not.toBeInTheDocument()

});
