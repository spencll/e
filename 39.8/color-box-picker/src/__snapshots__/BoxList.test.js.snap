// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Boxes
    </h3>
    <form>
      <label
        for="color"
      >
        Color: 
      </label>
      <input
        id="color"
        name="bgColor"
        placeholder="Box color"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height: 100
      </label>
      <input
        id="height"
        max="100"
        min="1"
        name="height"
        type="range"
        value="100"
      />
      <label
        for="width"
      >
        Width: 100
      </label>
      <input
        id="width"
        max="100"
        min="1"
        name="width"
        type="range"
        value="100"
      />
      <button>
        Add box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
